src/snackbar/index.js

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright (c) 2018-2020 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+
+// @flow
+
+import * as React from 'react';
+
+import {Button, KIND, SHAPE} from '../button/index.js';
+import {useStyletron} from '../styles/index.js';
+
+type PropsT = {|
+  actionMessage?: string,
+  actionOnClick?: (SyntheticEvent<HTMLButtonElement>) => mixed,
+  message: string,
+  startEnhancer?: React.AbstractComponent<{|
+    size: number,
+  |}>,
+|};
+
+const ActionButton = React.forwardRef(({onClick, message}, ref) => {
+  const [, theme] = useStyletron();
+  return (
+    <div>
+      <Button
+        ref={ref}
+        overrides={{
+          BaseButton: {
+            style: {
+              color: theme.colors.contentInversePrimary,
+              marginRight: '4px',
+              width: '100%',
+              ':hover': {
+                backgroundColor: theme.colors.borderInverseTransparent,
+              },
+              ':active': {
+                backgroundColor: theme.colors.backgroundInverseSecondary,
+              },
+            },
+          },
+        }}
+        kind={KIND.tertiary}
+        onClick={onClick}
+        shape={SHAPE.pill}
+      >
+        {message}
+      </Button>
+    </div>
+  );
+});
+
+export function Snackbar({
+  actionMessage,
+  actionOnClick,
+  message,
+  startEnhancer: StartEnhancer,
+}: PropsT) {
+  const [css, theme] = useStyletron();
+
+  const rootRef = React.useRef(null);
+  const [rootWidth, setRootWidth] = React.useState(0);
+  React.useEffect(() => {
+    function handleResize([entry]) {
+      setRootWidth(entry.contentRect.width);
+    }
+    const observer = new ResizeObserver(handleResize);
+    if (rootRef.current) {
+      observer.observe(rootRef.current);
+    }
+    return function cleanup() {
+      observer.disconnect();
+    };
+  }, []);
+
+  const actionMeasureRef = React.useRef(null);
+  const [actionMeasureWidth, setActionMeasureWidth] = React.useState(0);
+  React.useEffect(() => {
```

Comment:
I wonder if [`useLayoutEffect`](https://reactjs.org/docs/hooks-reference.html#uselayouteffect) would allow us to avoid the offscreen render?